### PR TITLE
Extend BlackBoxClassifier* to accept existing predictions

### LINKED ISSUE #1215: Extend BlackBoxClassifier* to accept a tuple of input/labels instead of callable
@abigailgold Thank you very much for this feature proposal.

We would like to run inference attacks without any access to a model and only a tuple of inputs and labels. Therefore we would like to extend `BlackBoxClassifier` and `BlackBoxClassifierNeuralNetwork` to accept also a tuple of inputs and labels and by default convert them into a predict method based on a look-up table.

### PR DESCRIPTION
**Pull request by GalHorowitz**
Signed-off-by: GalHorowitz <gal@galhorowitz.com>

# Description

Extend BlackBoxClassifier and BlackBoxClassifierNeuralNetwork to accept a tuple of inputs and labels and convert them into a predict method based on a look-up table.
Also supports using `np.isclose` for fuzzy matching inputs to prevent problems arising from floating-point inaccuracies.
To support fast lookups in the fuzzy case, the `sortedcontainers` dependency was added for a search-tree structure (`SortedList`), this is not needed in the normal case where a hash-table can be used.

Fixes #1215

## Type of change

Please check all relevant options.

- [x] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

All unit tests pass.

**Test Configuration**:
- OS: MacOS 11.4
- Python version:  3.8.11

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/estimators/classification/blackbox.py [hunk 1]
```diff
@@ -23,6 +23,8 @@
 import logging
 from typing import Callable, List, Optional, Union, Tuple, TYPE_CHECKING
 
+from functools import total_ordering
```
Comment (beat-buesser) [inline art/estimators/classification/blackbox.py:26]: Could you please move this import before `import logging` to group standard library imports in alphabetical order?

File: art/estimators/classification/blackbox.py [hunk 2]
```diff
@@ -287,3 +306,91 @@ def loss(self, x: np.ndarray, y: np.ndarray, **kwargs) -> np.ndarray:
 
     def compute_loss(self, x: np.ndarray, y: np.ndarray, **kwargs) -> np.ndarray:
         raise NotImplementedError
+
+
+@total_ordering
+class FuzzyMappingWrapper:
```
Comment (beat-buesser) [inline art/estimators/classification/blackbox.py:312]: Can we save the word Wrapper?
```suggestion
class FuzzyMapping:
```

File: art/estimators/classification/blackbox.py [hunk 3]
```diff
@@ -287,3 +306,91 @@ def loss(self, x: np.ndarray, y: np.ndarray, **kwargs) -> np.ndarray:
...
+    """
+    Wrapper class for a sample, label pair to be used in a `SortedList`.
```
Comment (beat-buesser) [inline art/estimators/classification/blackbox.py:314]:
```suggestion
    Class for a sample/label pair to be used in a `SortedList`.
```

File: art/estimators/classification/blackbox.py [hunk 4]
```diff
@@ -287,3 +306,91 @@ def loss(self, x: np.ndarray, y: np.ndarray, **kwargs) -> np.ndarray:
...
+    """
+
+    def __init__(self, key: np.ndarray, value=None):
+        """
+        Create a wrapper for the key, value to pair to be used in a `SortedList`.
```
Comment (beat-buesser) [inline art/estimators/classification/blackbox.py:319]:
```suggestion
        Create an instance of a key/value to pair to be used in a `SortedList`.
```

File: art/estimators/classification/blackbox.py [hunk 5]
```diff
@@ -287,3 +306,91 @@ def loss(self, x: np.ndarray, y: np.ndarray, **kwargs) -> np.ndarray:
...
+    :return: Prediction function.
+    """
+
+    samples, labels = existing_predictions
+
+    if fuzzy_float_compare:
+        from sortedcontainers import SortedList
+
+        # Construct a search-tree of the predictions, using fuzzy float comparison
+        sorted_predictions = SortedList([FuzzyMappingWrapper(key, value) for key, value in zip(samples, labels)])
```
Comment (beat-buesser) [inline art/estimators/classification/blackbox.py:363]:
```suggestion
        sorted_predictions = SortedList([FuzzyMapping(key, value) for key, value in zip(samples, labels)])
```

File: art/estimators/classification/blackbox.py [hunk 6]
```diff
@@ -287,3 +306,91 @@ def loss(self, x: np.ndarray, y: np.ndarray, **kwargs) -> np.ndarray:
...
+
+        def fuzzy_predict_fn(batch):
+            predictions = []
+            for row in batch:
+                try:
+                    match_idx = sorted_predictions.index(FuzzyMappingWrapper(row))
```
Comment (beat-buesser) [inline art/estimators/classification/blackbox.py:369]:
```suggestion
                    match_idx = sorted_predictions.index(FuzzyMapping(row))
```

File: art/estimators/classification/blackbox.py [hunk 7]
```diff
@@ -287,3 +306,91 @@ def loss(self, x: np.ndarray, y: np.ndarray, **kwargs) -> np.ndarray:
...
+        self.key = key
+        self.value = value
+
+    def __eq__(self, other):
+        return np.all(np.isclose(self.key, other.key))
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def __ge__(self, other):
```
Comment (beat-buesser) [inline art/estimators/classification/blackbox.py:333]: I haven't used `@total_ordering` before and looked it up in the docs and noticed the note that it comes at the cost of slower execution. You have done performance tests with `SortedList`, does `SortedList` only use `__ge__` or could explicitly defining all comparison magic functions further improve performance here?

File: tests/estimators/classification/test_blackbox_existing_predictions.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+import pytest
+
+import numpy as np
+
+from art.estimators.classification import BlackBoxClassifierNeuralNetwork, BlackBoxClassifier
```
Comment (beat-buesser) [inline tests/estimators/classification/test_blackbox_existing_predictions.py:22]:
```suggestion
from art.estimators.classification.blackbox import BlackBoxClassifierNeuralNetwork, BlackBoxClassifier
```

**Review by beat-buesser — CHANGES_REQUESTED** [review]
Hi @GalHorowitz Thank you very much for your first pull request to ART enabling attacks on the privacy of data and predictions without the requirement of direct access to the actual model!

I think the pull request looks great and I have only made comments in a few places where I thought that maybe small improvements could be possible.

There is a `notebooks/README.md` which contains links to nbviewer and short abstracts for each notebook, could you please add your new notebook?

Maybe, you could highlight again in the title or description of the last cell of the notebook that this attack is now running without accessing the model because this is quite unique and novel for ART.

File: art/estimators/classification/blackbox.py (hunk 7 above)
Comment (GalHorowitz) [reply art/estimators/classification/blackbox.py:333]: It seems `SortedList` uses python's `bisect`, which does use all comparison types. I ran the tests using `@total_ordering` and without, and the performance difference seems negligible (on the order of ms). But I did notice that by not calling `__eq__` in `__ge__` a repeated call to `np.isclose` can be removed, so I was able to halve the running time.

**Comment by GalHorowitz** [comment]
@beat-buesser, thank you for the review.
Regarding `notebooks/README.md`, it seems that nbviewer links all link to the main branch, but this PR is for the `dev_1.8.0` branch. Should I link to this branch, or link to the `master` branch (which will not have this notebook, for now?)

File: art/estimators/classification/blackbox.py (hunk 7 above)
Comment (beat-buesser) [reply art/estimators/classification/blackbox.py:333]: That's great news!

**Comment by beat-buesser** [comment]
@GalHorowitz Please prepare the nbviewer links to link branch `main` even if the PR first goes to branch `dev_1.8.0`.

**Review by beat-buesser — APPROVED** [review]
Hi @GalHorowitz Thank you very much! This looks good to me.

**Pull request merged**